keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,393 @@
+- Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
+not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. 
+
+Having the above checks in kubelet helps in the following conditions:
+- If scheduler has been by passed by setting the nodeName directly on the pod spec.
+- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.
+
+As of now, Kubelet doesn't reconcile the value of `kubernetes.io/os`. So, it'd be nice to have the reconciliation as part of this change.
+
+
+### Implications to Scheduler
+
+We let the users to explicitly specify nodeSelectors+tolerations or runtimeclasses to express their intention
+to run an particular OS. However, in future, once the OS struct expands, we can see if we can leverage those fields to
+express scheduling constraints. During the alpha, we assume there are no scheduling implications.
+
+
+### Test Plan
+
+- Unit tests covering API server defaulting to various fields within pod with and without this feature
+- Unit tests covering admission plugins which validate/mutate pod spec based on this feature
+- Unit and E2E tests for Kubelet changes.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Basic units and e2e tests completed and enabled
+
+#### Alpha -> Beta Graduation
+- Expand the e2e tests with more scenarios
+- Gather feedback from end users
+- Tests are in Testgrid and linked in the KEP
+
+#### Alpha -> Beta Graduation
+- 2 examples of end users using this field
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades:
+  When upgrading from a release without this feature, to a release with `IdentifyOS` feature enabled, there is no change to existing behavior unless user specifies
+  OS field in the pod spec.
+- Downgrades:
+  When downgrading from a release with this feature to a release without `IdentifyOS`, the current behavior will continue.
+### Version Skew Strategy
+
+If the feature gate is enabled there are some kubelet implications as the code to strip security constraints based on OS can be removed and we need to add
+admission/denying in the kubelet logic which was mentioned above. Older Kubelets without this change will continue stripping the unnecessary fields in the pod spec which is the current behavior.
+
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: IdentifyOS
+  - Components depending on the feature gate:
+    - kubelet
+    - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+No
```

Comment:
```suggestion
A Kubelet with a misscheduled pod (i.e. trying to run a windows pod on a linux node) will fail *before* trying to run a container, as opposed to after.
```

Reply:
(@ravisantoshgudimetla keep me honest here in case im missing something)

Reply:
> (@ravisantoshgudimetla keep me honest here in case im missing something)

That's how I understand the kubelet change 

> Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
not matching `runtime.GOOS`